### PR TITLE
Preserve QOBLIB source provenance

### PR DESCRIPTION
**Pull request by bernalde**
Summary

- Adds optional `source_metadata` persistence for HDF5 source groups so source provenance attributes live under `/instances/{id}/source`.
- Adds forward-compatible QOBLIB ingestion for canonical QS/QUBO artifacts when a group declares authoritative LP source metadata: it stores LP provenance and bounded source text, and populates `source_objective`, `dual_bound`, and `source_feasible` when source evaluation is possible. Current shipped QOBLIB groups do not declare source model paths, so this path is covered by fixtures and future source-aware groups.
- Documents the QOBLIB source-text storage policy, clarifies `dual_bound` as the canonical imported bound with `objective_bound` retained for compatibility, and updates source-schema release notes.

Tests

- `julia --project -e 'import Pkg; Pkg.test()'` passed, 149 tests.
- `make test-build` passed, 176 tests.
- `julia --project=docs ./docs/make.jl --skip-deploy` passed.
- `git diff --check` passed.

Branch Hygiene

- Base branch: `main`.
- Source branch: `fix/issue-46-source-formulations`.
- Source branch point: current `origin/main` at `ef0f24b`.
- Stacked status: not stacked.
- Prerequisite PRs: none. The previous PR from this branch, #48, is already merged into `main`; this PR contains one new commit on top of current `main`.

Refs #46

### CODE REVIEW

File: scripts/build/sources/qoblib.jl [hunk 1]
```diff
@@ -848,6 +849,159 @@ function _qoblib_source_path(root_path::AbstractString, path::AbstractString)
...
 
+function _qoblib_group_source_format(group)
+    if hasproperty(group, :source_model_format)
+        return lowercase(String(group.source_model_format))
+    else
+        return "lp"
+    end
+end
+
+function _qoblib_group_source_roots(root_path::AbstractString, group)
```
Comment (bernalde) [inline scripts/build/sources/qoblib.jl:860]: Question: this and `_qoblib_source_model_path` only fire when a group declares `source_model_path`/`source_model_paths`. None of the four real `QOBLIB_GROUPS` do, so for the canonical build this returns empty and no source group is written for any shipped instance — the whole source-ingestion path is exercised only by the `constrainedfixture` test group. Is a real class meant to be wired here, or is this intentional scaffolding for a future group? If the latter, the PR description's claim that it stores source for canonical QS artifacts overstates the current effect.

File: scripts/build/sources/qoblib.jl [hunk 2]
```diff
@@ -848,6 +849,159 @@ function _qoblib_source_path(root_path::AbstractString, path::AbstractString)
...
+            source_text = nothing,
+            source_encoding = nothing,
+            source_metadata = nothing,
+        )
+    end
+
+    source_bytes = read(path)
+    source_format = _qoblib_group_source_format(group)
+    limit = _qoblib_source_text_max_bytes(group)
+    store_text = length(source_bytes) <= limit
```
Comment (bernalde) [inline scripts/build/sources/qoblib.jl:965]: Nonblocking: the `store_text = false` branch (source over the byte limit -> `content` omitted, `source_text_storage = "omitted_size_limit"`, `source_text = nothing` but provenance/hash retained) is the headline policy of this PR and is untested. The fixture only hits the `"stored"` path (runtests.jl:771). Add a fixture group with a small `source_text_max_bytes` and a larger source file so the omission branch is validated.

File: scripts/build/sources/qoblib.jl [hunk 3]
```diff
@@ -848,6 +849,159 @@ function _qoblib_source_path(root_path::AbstractString, path::AbstractString)
...
+
+    return (
+        source_format = source_format,
+        source_text = store_text ? String(source_bytes) : nothing,
+        source_encoding = _qoblib_source_encoding(root_path, group, model_path),
+        source_metadata = metadata,
+    )
+end
+
+function _qoblib_source_evaluation(index::QUBOLib.LibraryIndex, instance::Integer, state)
```
Comment (bernalde) [inline scripts/build/sources/qoblib.jl]: Nonblocking (two points): (1) Performance - this calls `evaluate_source`, which re-parses the LP via `source_model` (mktempdir + `JuMP.read_from_file`) and re-reads the encoding via `project_solution` on every submission and incumbent. For an instance with many records the same LP is parsed once per record; parse the model and encoding once per instance and reuse. (2) Robustness - `evaluate_source` is unguarded here, so a bad encoding or a state/dimension mismatch aborts the entire build. The existing source_value-disagreement code logs and continues; consider catching, recording missing, and warning for consistency.

File: scripts/build/sources/qoblib.jl [hunk 4]
```diff
@@ -1360,7 +1517,10 @@ function _add_qoblib_submission!(
             submission,
             qubo_value,
             source_value = record_source_value,
+            source_objective = source_evaluation.source_objective,
             objective_bound,
+            dual_bound = objective_bound,
```
Comment (bernalde) [inline scripts/build/sources/qoblib.jl:1607]: Question: `objective_bound` comes from the QOBLIB "Optimality Bound" column (line 1419), so populating `dual_bound` from it is defensible, but the identical value now lands in both `objective_bound` and `dual_bound` for submission/unavailable/unmapped records (incumbents get neither). Is the duplication intentional? If `dual_bound` is the canonical field per #46, note that `objective_bound` is kept for back-compat, or drop one, so consumers know which column is authoritative.

**Review by bernalde — COMMENTED** [review]
Review as maintainer. Note: I am the PR author, so GitHub only permits a COMMENT event here; the formal verdict (approval or change request) must come from another maintainer. Based on the findings below there are no blocking issues.

Setup: fetched origin/main and pull/50/head (a76daed); merge-base ef0f24b matches the stated branch point. Diff vs merge-base matches the reported totals (352 additions, 10 deletions, 9 files). Reviewed at head in a clean tree.

Scope and intent: correctly linked Refs #46 (not Closes), which is right — this is incremental provenance plumbing on top of the #48 source APIs, not a full resolution of #46. It adds `source_metadata` persistence to `add_instance!`/`_write_instance_source!`, QOBLIB ingestion wiring to discover LP source models and store provenance + bounded source text, and populates `source_objective`/`dual_bound`/`source_feasible`. The library-side `source_metadata` and the QOBLIB fixture path are tested.

One thing to flag about the description's claim that it "wires QOBLIB ingestion to discover matching LP source models for canonical QS artifacts": I verified that none of the four real `QOBLIB_GROUPS` (marketsplit, labs, portfolio, independentset) declare `source_model_path(s)` or `source_encoding`, so on the actual canonical build `_qoblib_instance_source` returns all-nothing and no source group is written for any real instance; `_qoblib_source_evaluation` likewise returns missing. The feature currently executes only against the synthetic `constrainedfixture` test group. That is consistent with reality (the unconstrained classes have no LP source; the constrained classes have no QS artifacts per PR #49's inventory), so it reads as forward-looking scaffolding — acceptable under Refs #46, but see Question 1.

1. Blocking issues: none found.

2. Nonblocking issues:
- scripts/build/sources/qoblib.jl: the >1MB size-limit branch (`store_text = false`, `content` omitted, `source_text_storage = "omitted_size_limit"`, `source_text = nothing`) is not covered by any test; the fixture only exercises the `"stored"` path. Add a fixture whose source exceeds its configured limit (a group can set `source_text_max_bytes`) so the omission branch and the no-`content`/still-has-provenance behavior are validated.
- scripts/build/sources/qoblib.jl `_qoblib_source_evaluation`: re-derives the source model through `evaluate_source` -> `source_model` (mktempdir + write + `JuMP.read_from_file`) and `project_solution` (re-reads the encoding) for every submission and every incumbent. When a group does carry source, an instance with many records re-parses the same LP once per record. Consider parsing the JuMP model and encoding once per instance and reusing them.
- scripts/build/sources/qoblib.jl `_qoblib_source_evaluation`: the `evaluate_source` call is unguarded. A malformed encoding or a state/dimension mismatch would raise and abort the whole build, unlike the existing source_value-disagreement handling that logs via `@info` and continues. Consider catching, recording missing, and warning, for consistency.

3. Questions:
- Inert canonical wiring: is a real group meant to be wired to source models in this PR, or is this intentionally scaffolding for a future class? As-is, nothing in the canonical artifact gains a source group, so the user-visible effect on the shipped artifact is nil. If intentional, consider saying so in the PR description to avoid implying the canonical build now stores QOBLIB source.
- `dual_bound = objective_bound`: `objective_bound` is sourced from the QOBLIB submission CSV "Optimality Bound" (line 1419), so mapping it to `dual_bound` is defensible, but the same value is now written to two columns (`objective_bound` and `dual_bound`) for submission/unavailable/unmapped records, while incumbents get neither. Is the duplication intentional? If `dual_bound` is the canonical field going forward, document that `objective_bound` is retained for back-compat, or stop writing one of them, so downstream consumers know which to trust.

4. Tests run and outcomes (Julia 1.10.11, clean checkout at head):
- `julia --project=scripts/build -e 'include("scripts/build/runtests.jl"); test_main()'` - 160/160 pass, including the expanded 70-test "QOBLIB build fixture" set.
- `julia --project -e 'import Pkg; Pkg.test()'` - 147/147 pass; "Source formulations" at 21, "Website docs" at 60. The PR's reported counts are accurate.
- Full Documenter HTML build not rerun (needs docs Manifest/network); make.jl wiring and docs.jl assertions are covered by the package suite.

5. Merge-readiness: tests green, no API break, backward compatible (new kwargs default to nothing; `source_metadata` is additive and the schema's `metadata`/source group are open). The two Questions and the missing size-limit test are worth resolving before merge, but none block. No blocking issues, so I would not withhold merge on the findings above; the formal approval must come from another maintainer since I authored this PR.

File: scripts/build/sources/qoblib.jl (hunk 1 above)
Comment (bernalde) [reply scripts/build/sources/qoblib.jl:860]: Addressed with a PR body update. It now states that source ingestion is forward-compatible fixture coverage unless a QOBLIB group declares source model paths; current shipped groups still do not. No code commit was needed for this thread.

File: scripts/build/sources/qoblib.jl (hunk 2 above)
Comment (bernalde) [reply scripts/build/sources/qoblib.jl:965]: Addressed in https://github.com/JuliaQUBO/QUBOLib.jl/commit/9065751e708c034a0f385f59a87978a23eaf2683. Added `limitedsourcefixture` with `source_text_max_bytes = 16`; tests assert `content` is omitted, `source_text_storage = "omitted_size_limit"`, and provenance/hash/encoding are retained in `scripts/build/runtests.jl`.

File: scripts/build/sources/qoblib.jl (hunk 3 above)
Comment (bernalde) [reply scripts/build/sources/qoblib.jl]: Addressed in https://github.com/JuliaQUBO/QUBOLib.jl/commit/9065751e708c034a0f385f59a87978a23eaf2683. The build now creates one source evaluator per instance, reuses the parsed LP model and encoding for records, and catches evaluation failures to warn and record missing source fields. Added malformed-encoding coverage in `scripts/build/runtests.jl`.

File: scripts/build/sources/qoblib.jl (hunk 4 above)
Comment (bernalde) [reply scripts/build/sources/qoblib.jl:1607]: Addressed in https://github.com/JuliaQUBO/QUBOLib.jl/commit/9065751e708c034a0f385f59a87978a23eaf2683. Documented `dual_bound` as the canonical imported QOBLIB bound and `objective_bound` as the backward-compatible copy in `docs/src/manual/3-qoblib.md` and `RELEASE.md`.
